pkg/ottl/expression.go

```diff
@@ -549,6 +550,89 @@ func (g StandardIntLikeGetter[K]) Get(ctx context.Context, tCtx K) (*int64, erro
 	return &result, nil
 }
 
+// ByteSliceLikeGetter is a Getter that returns []byte by converting the underlying value to an []byte if necessary
+type ByteSliceLikeGetter[K any] interface {
+	// Get retrieves []byte value.
+	// The expectation is that the underlying value is converted to []byte if possible.
+	// If the value cannot be converted to []byte, nil and an error are returned.
+	// If the value is nil, nil is returned without an error.
+	Get(ctx context.Context, tCtx K) (byteSlice, error)
+}
+
+type StandardByteSliceLikeGetter[K any] struct {
+	Getter func(ctx context.Context, tCtx K) (any, error)
+}
+
+//nolint:revive
+func (g StandardByteSliceLikeGetter[K]) Get(ctx context.Context, tCtx K) (byteSlice, error) {
+	val, err := g.Getter(ctx, tCtx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting value in %T: %w", g, err)
+	}
+	if val == nil {
+		return nil, nil
+	}
+	var result byteSlice
+	switch v := val.(type) {
+	case byteSlice:
+		result = v
+	case []byte:
+		result = v
+	case string:
+		result = []byte(v)
+	case float64:
+		result = float64ToBytes(v)
+	case int64:
+		result = int64ToBytes(v)
+	case bool:
+		if v {
+			result = []byte("1")
+		} else {
+			result = []byte("0")
+		}
+	case pcommon.Value:
+		switch v.Type() {
+		case pcommon.ValueTypeBytes:
+			result = v.Bytes().AsRaw()
+		case pcommon.ValueTypeInt:
+			result = int64ToBytes(v.Int())
+		case pcommon.ValueTypeDouble:
+			result = float64ToBytes(v.Double())
+		case pcommon.ValueTypeStr:
+			result = []byte(v.Str())
+		case pcommon.ValueTypeBool:
+			if v.Bool() {
+				result = []byte("1")
+			} else {
+				result = []byte("0")
+			}
+		default:
+			return nil, TypeError(fmt.Sprintf("unsupported value type: %v", v.Type()))
+		}
+	default:
+		return nil, TypeError(fmt.Sprintf("unsupported type: %T", v))
+	}
+	return result, nil
+}
+
+func float64ToBytes(f float64) []byte {
```

Comment:
I'm honestly not entirely clear on what's going on here without digging deeper into how floats are represented as bytes, could you put a comment on this function describing which parts of the float are encoded into the byte slice? Do we include the fractional part, or do we truncate it?

A comment on `ByteSliceGetter` to describe how floats are handled would also work.

Reply:
Thanks for adding the comments. My confusion arose from missing that `64 - 8 = 56` and not being sure if we were omitting part of the float byte representation.

I think the code is easy to understand if we explain the assumed endianness of the value and that we take the whole thing, but others may not be as familiar with bit manipulation, so I'm okay being verbose.